### PR TITLE
fix: More compatibility fixes for Terraform v0.13 and aws v3

### PR DESCRIPTION
**Pull request by dpiddockcmp**
# PR o'clock

## Description

Terraform v0.13 and aws v3 support time!

- The update to the vpc module was, strictly speaking, unnecessary but it adds the `terraform` block with supported versions.
- Update for iam module was very necessary to support new versions
- Workaround for "Provider produced inconsistent final plan" when creating ASGs at the same time as the cluster. See  https://github.com/terraform-providers/terraform-provider-aws/issues/14085 for full details.

When upgrading an existing TF 0.12/aws 2 cluster to TF 0.13/aws 3 you currently have to manually edit the state file to delete removed attributes. Should be fixed in v0.13.1 https://github.com/hashicorp/terraform/pull/25779 . For the basic example I had to remove the following attributes:
- data `aws_availability_zones`: `blacklisted_names` and `blackedlisted_zone_ids`
- `aws_iam_instance_profile`: `roles` (#974)

You also need to mess around with `replace-provider` a lot:
```sh
terraform state replace-provider -- -/aws hashicorp/aws
terraform state replace-provider -- -/kubernetes hashicorp/kubernetes
terraform state replace-provider -- -/local hashicorp/local
terraform state replace-provider -- -/null hashicorp/null
terraform state replace-provider -- -/random hashicorp/random
terraform state replace-provider -- -/template hashicorp/template
```

### Checklist

- [ ] CI tests are passing
- [ ] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

### CODE REVIEW

File: workers.tf [hunk 1]
```diff
@@ -102,31 +102,38 @@ resource "aws_autoscaling_group" "workers" {
...
-        "value"               = "owned"
-        "propagate_at_launch" = true
-      },
-    ],
-    local.asg_tags,
-    lookup(
-      var.worker_groups[count.index],
-      "tags",
-      local.workers_group_defaults["tags"]
+  dynamic "tag" {
```
Comment (barryib) [inline workers.tf:105]: I don't test it yet, but is there any reason to change `tags` into `dynamic "tag"` ?

Comment (barryib) [reply workers.tf:105]: Never mind. I read https://github.com/terraform-providers/terraform-provider-aws/issues/14085

**Comment by daroga0002** [comment]
@dpiddockcmp  looks Terraform 13.1 was released so should solve mentioned bug

https://github.com/hashicorp/terraform/releases/tag/v0.13.1

**Comment by dpiddockcmp** [comment]
@barryib are we good to merge this?

**Comment by barryib** [comment]
Please let me test this tonight with TF 0.13.1

**Review by barryib — APPROVED** [review]

**Comment by barryib** [comment]
@dpiddockcmp LGTM. Just tested an upgrade TF 0.12 to TF 0.13.1 without error. Shouldn't we change the Terraform required version to avoid TF 0.13.0 (`"~> 0.12.9, >= 0.13.1"`) ?

**Comment by dpiddockcmp** [comment]
That version constraint can't be met. It can't be `0.12.*` and greater than `0.13.0`. Could use `>= 0.12.9, != 0.13.0`

**Comment by barryib** [comment]
> That version constraint can't be met. It can't be `0.12.*` and greater than `0.13.0`. Could use `>= 0.12.9, != 0.13.0`

Ok. LGTM.

**Review by barryib — APPROVED** [review]

**Pull request merged**